WOM (only!!) updates to add directories and additional files.

kshakir opened the pull request:
Updates to WomFile equals/add.
Future work will add CWL parsing and cromwell generation/usage.
Updates to other parts of code to support changes to WOM.
IntelliJ recommended updates to use lengthCompare.

cjllanwarne commented on wom/src/main/scala/wom/values/WomFile.scala:
```diff
@@ -1,67 +1,154 @@
+  * @param contentsOption The optional text contents of the file.
+  * @param secondaryFiles Any files associated with this file.
+  */
+case class WomSingleFile(value: String,
```
`final case class`

cjllanwarne commented on engine/src/main/scala/cromwell/engine/WdlFunctions.scala:
```diff
@@ -2,18 +2,20 @@ package cromwell.engine
   // Cromwell does not support writing files from the engine.
-  override def writeFile(path: String, content: String): Future[WomFile] = Future.failed(new Exception("Can't write files"))
+  override def writeFile(path: String, content: String): Future[WomSingleFile] = {
+    Future.failed(new Exception("Can't write files"))
```
ToL: while you're here, the comment 2 lines up is way more informative than the error text we're returning to users

cjllanwarne commented on supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesAsyncBackendJobExecutionActor.scala:
```diff
@@ -221,6 +221,9 @@ class JesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
 
     val outputs = womFileOutputs.distinct flatMap { womFile =>
       womFile match {
+        case _: WomSingleDirectory =>
```
Are you foreseeing a `WomMultiDirectory`?

cjllanwarne commented on wdl/src/main/scala/wdl/AstTools.scala:
```diff
@@ -112,6 +112,7 @@ object AstTools {
         case t: Terminal =>
           t.getSourceString match {
             case WomSingleFileType.`toDisplayString` => WomSingleFileType
+            case WomSingleDirectoryType.`toDisplayString` => WomSingleDirectoryType
```
ToL: Half way to supporting `Directory` in WDL too? 😎

cjllanwarne commented on wdl/src/main/scala/wdl/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -98,15 +97,17 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
       case Success(f) :: Nil => Success((f, WomString("")))
       case Success(f) :: Success(s) :: Nil => Success((f, s))
-      case s if s.size > 2 || s.size < 1 => Failure(new IllegalArgumentException(s"Bad number of arguments to basename(filename, suffixToStrip = ''): ${params.size}"))
+      case s if s.lengthCompare(2) > 0 || s.lengthCompare(1) < 0 =>
```
ToL: it might be "better" scala but IMO their new syntax is way less readable

kshakir replied:
> new syntax is way less readable

That should be scala's official motto 😉

cjllanwarne commented on wom/src/main/scala/wom/types/WomFileType.scala:
```diff
@@ -1,56 +1,69 @@
+sealed trait WomSingleDirectoryOrFileType extends WomFileType
+
+case object WomSingleDirectoryType extends WomSingleDirectoryOrFileType {
+  val toDisplayString: String = "Dir"
```
ToL: can we please use `"Directory"` instead? (!!)

cjllanwarne commented on wom/src/main/scala/wom/types/WomPrimitiveType.scala:
```diff
@@ -1,11 +1,13 @@
     WomStringType -> Seq(WomStringType, WomIntegerType, WomFloatType, WomSingleFileType, WomBooleanType),
+    WomSingleDirectoryType -> Seq(WomStringType, WomSingleDirectoryType),
     WomSingleFileType -> Seq(WomStringType, WomSingleFileType),
+    WomGlobFileType -> Seq(WomStringType, WomGlobFileType),
```
Heads up: `WomGlobFile`s are only ever returned by `evaluateFiles` as a placeholder and never actually passed around in Cromwell. 
- When PAPI2 removes the need for `evaluateFiles`, we should be able to remove `WomGlobFile` too
- If we're trying to coerce to or from them, or store them in any value field in memory or in the database, we're probably doing something wrong.

cjllanwarne commented on wom/src/main/scala/wom/types/WomStringType.scala:
```diff
@@ -23,7 +23,8 @@ case object WomStringType extends WomPrimitiveType {
 
   override def add(rhs: WomType): Try[WomType] = rhs match {
-    case WomStringType | WomIntegerType | WomFloatType | WomSingleFileType => Success(WomStringType)
+    case WomStringType | WomIntegerType | WomFloatType | WomSingleDirectoryType | WomSingleFileType | WomGlobFileType =>
```
IMO `String + File` doesn't really make sense and should be outlawed

cjllanwarne commented on wom/src/main/scala/wom/util/JsUtil.scala:
```diff
@@ -24,7 +24,7 @@ object JsUtil {
     * The WomMap keys and values, and WomArray elements must be the one of the above, recursively.
     *
-    * WomSingleFile and WomGlobFile are not permitted, and must be already converted to one of the above types.
+    * WomFile are not permitted, and must be already converted to one of the above types.
```
Grammar are wrong here

cjllanwarne commented on wom/src/main/scala/wom/values/WomFile.scala:
```diff
@@ -1,67 +1,154 @@
+                         sizeOption: Option[Long] = None,
+                         formatOption: Option[String] = None,
+                         contentsOption: Option[String] = None,
+                         secondaryFiles: Seq[WomSingleDirectoryOrFile] = Vector.empty
```
ToL: Is it just me or is `secondaryFiles` an odd field for `WomSingleFile` to have? ie we can no longer make assumptions like "this is a single file"

Horneth replied:
Yeah I agree this feels a bit weird but I don't have a better suggestion

cjllanwarne commented on wom/src/test/scala/wom/types/WomFileTypeSpec.scala:
```diff
@@ -0,0 +1,124 @@
+    ("a wom glob to a glob", WomGlobFileType, WomGlobFile("example/glob*"), WomGlobFile("example/glob*"))
+  )
+
+  forAll(coercionTests) { (description, womFileType, value, expected) =>
```
An observation: you could remove the `womFileType` field and just use `expected.womType` to make these test cases a little shorter

cjllanwarne commented on wom/src/test/scala/wom/types/WomFileTypeSpec.scala:
```diff
@@ -0,0 +1,124 @@
+    ("a wom file to a dir", WomSingleFileType, WomSingleDirectory("example/dir"),
+      """No coercion defined from '"example/dir"' of type 'Dir' to 'File'."""),
+    ("a wom file to a glob", WomSingleFileType, WomGlobFile("example/glob*"),
+      """No coercion defined from 'glob("example/glob*")' of type 'Glob' to 'File'."""),
```
The description says `File` to `Glob` but the error says `"...Glob to File"`

cjllanwarne replied:
Actually that applies to all these cases

cjllanwarne commented on wom/src/test/scala/wom/types/WomFileTypeSpec.scala:
```diff
@@ -0,0 +1,124 @@
+
+  forAll(failedCoercionTests) { (description, womFileType, value, expected) =>
+    it should s"fail to coerce $description" in {
+      womFileType.coerceRawValue(value).failed.get.getMessage should be(expected)
```
I believe there's a `should throwWithExceptionMessage()` or some such in scalatest

Horneth commented on wom/src/main/scala/wom/types/WomFileType.scala:
```diff
@@ -1,56 +1,69 @@
+  override def equals(rhs: WomType): Try[WomType] = rhs match {
+    case _: WomFileType => Success(WomBooleanType)
+    case WomStringType => Success(WomBooleanType)
+    case WomOptionalType(memberType) => equals(memberType)
```
Wouldn't this say that `WomFileType equals WomOptionalType(WomFileType)` ? Unless that's the intended behaviour ?

cjllanwarne replied:
This lets you say:
```wdl
Int? input
if (input == 5) { ... }
```

Horneth commented on wom/src/main/scala/wom/values/WomFile.scala:
```diff
@@ -1,67 +1,154 @@
 
 sealed trait WomFile extends WomPrimitive {
+
+  // TODO: WOM: WOMFILE: This is often used as _either_ the cloud location _or_ the path within the container. Does this need to be split into two values?
```
In a long and distant future something like a `WomPath` could be interesting I think

danbills commented on wom/src/main/scala/wom/util/JsUtil.scala:
```diff
@@ -72,6 +72,25 @@ object JsUtil {
+    *
+    * @param value A WOM value.
+    * @return The javascript equivalent.
+    */
```
💯

danbills commented on wom/src/main/scala/wom/values/WomFile.scala:
```diff
@@ -1,67 +1,154 @@
+    * @param f The function to update the location.
+    * @return A new WomFile with the updated location.
+    */
+  def mapFile(f: String => String): WomFile
```
is it full path or the filename being passed in here?

kshakir replied:
I can't remember off the top of my head. This might even be the relative path that gets converted to a full path by the backend.

danbills replied:
* What does "convert" mean?  I.e. from what to what?
 * Is there any way we could use types to better define what is going in and out?

kshakir commented:
PR comments addressed, and tests re-passing. Changes include:

- Primitive and non-primitive file types.
- Primitive 'Dir' now 'Directory'.
- Split up and fixed test descriptions.
- Exception patches.

danbills commented on wom/src/main/scala/wom/values/WomFile.scala:
```diff
@@ -1,67 +1,207 @@
+  * Should not be passed into command line generation. Instead, the execution engine should create a WomListedDirectory
+  * locating the files/directories within the `value` and filling in the listing.
+  *
+  * @param value The location of the directory, possibly in the cloud.
```
could we rename `value` to `location` or `path` or similar?

kshakir replied:
I'm leaving it as is for this PR. It was previously called `WomFile.value` and changing it would add 113 lines of diff to sort through. Should be a very quick change for IntelliJ in a future renaming-PR.
